### PR TITLE
[Feat] Gpu data filtering - 1

### PR DESCRIPTION
**Pull request by heshan0131**
Step. 1 GPU Data filtering feature (https://github.com/uber/kepler.gl/issues/328)
- Add `gpu` property to filter. 
- Implement logic to manage which filter is gpu enabled
- Remove filtered data array from filter result (not used anywhere)
- Implement `set` function to immutably edit reducer state

### CODE REVIEW

File: src/utils/layer-utils/layer-utils.js [hunk 1]
```diff
@@ -48,24 +48,24 @@ export function findDefaultLayer(dataset, layerClasses) {
...
- * @param {object} layer
- * @param {object} state
- * @param {object} oldLayerData
- * @param {object} opt
- * @returns {object} {layerData: {}, layer: {} || undefined}
+ * @param {Object} layer
+ * @param {Object} state
+ * @param {Object} oldLayerData
+ * @param {Object} opt
+ * @returns {Object} {layerData: {}, layer: {} || undefined}
```
Comment (macrigiuseppe) [inline src/utils/layer-utils/layer-utils.js]: i noticed you are using two different annotation styles to document object types:
1. {Object} {layerData: {}, layer: {} || undefined}
2 {{layerData: {}, layer: {} || undefined}}

We should pick one and be consistent

File: src/utils/filter-utils.js [hunk 2]
```diff
@@ -246,55 +338,43 @@ export function filterData(data, dataId, filters) {
...
+  // generate 2 sets of
+  // filtered index used to calculate layer data and layer Domain
+  const filteredIndex = [];
+  const filteredIndexForDomain = [];
 
-        // filter data for render
-        const matchForRender = fixedDomainFilters.every(filter =>
-          isDataMatchFilter(d, filter, i)
-        );
+  for (let i = 0; i < allData.length; i++) {
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/utils/filter-utils.js]: js benchmark proves that for loop is 30% faster than map / reduce. deck.gl have been using for loop in almost all data processing functions

File: src/utils/utils.js [hunk 3]
```diff
@@ -59,5 +59,39 @@ export function notNullorUndefined(d) {
 }
 
 export function capitalizeFirstLetter(str) {
-  return str.charAt(0).toUpperCase() + str.slice(1);
+  return typeof str === 'string' ? str.charAt(0).toUpperCase() + str.slice(1) : str;
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/utils/utils.js]: Previously we have a couple of crashes in this function then some random label is null. I would rather avoid it

File: src/utils/filter-utils.js [hunk 4]
```diff
@@ -59,6 +59,8 @@ export const PLOT_TYPES = keyMirror({
   lineChart: null
 });
 
+export const MAX_GPU_FILTERS = 4;
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/utils/filter-utils.js]: This is the max amount of buffer for filter value deck.gl allows

[the remark this answers is not in the document]
Comment (macrigiuseppe) [reply src/utils/filter-utils.js]: I would add a comment to make sure people understand this is is something we are not in control of

File: src/utils/utils.js [hunk 5]
```diff
@@ -59,5 +59,39 @@ export function notNullorUndefined(d) {
...
 }
+
+/**
+ * immutably insert value to an Array or Object
+ * @param {Array|Object} obj
+ * @param {Number|String} key
+ * @param {*} value
+ * @returns {Array|Object}
+ */
+const insertValue = (obj, key, value) => {
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/utils/utils.js]: yes we can do that.
But for the leaves inside an object, the idea is to create it
```js
set(['map'], 'hello', null) // null

set(['map'], 'hello', {}) // {map: 'hello')
```

**Comment by heshan0131** [comment]
renamed branch to `gpu-data-filter-1`
https://github.com/uber/kepler.gl/pull/451

**Pull request closed without merging**